1-Draft/RFCNNNN-#Requires-Additions.md

```diff
@@ -0,0 +1,155 @@
+---
+RFC: RFC
+Author: Robert Holt
+Status: Draft
+SupercededBy:
+Version:
+Area: Hash-requires
+Comments Due: 2018-05-15
+Plan to implement: Yes
+---
+
+# \#Requires Additions
+
+Currently, PowerShell's `#requires` statement (or perhaps
+pragma?) supports the following parameters:
+
+* `-Version <N>[.<n>]`, where a minimum PowerShell version can be specified
+* `-PSSnapin <PSSnapin-Name> [-Version <N><n>]`, where a required PowerShell Snapin can be specified
+* `-Modules { <Module-name> | <Hashtable> }`, where PowerShell modules that are required can be specified
+* `-ShellId <ShellId>`, where the required Shell ID can be specified
+* `-RunAsAdministrator`, where the script is required to be run as administrator
+
+These features are documented in [about_Requires](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-6), along with the statement that:
+
+> You can use #Requires statements in any script. You
+> cannot use them in functions, cmdlets, or snap-ins.
+
+Currently however, this is untrue, as `#requires`
+statements are allowed by the parser/tokenizer anywhere in
+a script and then effectively hoisted to the top of that
+script to be checked before any part of the script is
+executed, no matter where they are placed.
+
+This RFC proposes the following changes:
+
+* Only allow `#requires` at the top level of a script,
+  before any lines that are not comments (i.e. with the
+  intention that a hashbang can still work, just before
+  any executable PowerShell code). Placing `#requires` anywhere
+  after will cause a parse-time error. This would be a **breaking
+  change**, albeit one that the documentation already claims to be 
+  in force.
+* Using `#requires` in the interactive console will cause
```

Comment:
I'm not sure this provides any benefit, and I'm not aware of a straightforward way to implement it.

Today the engine knows nothing about interactive input - that's the domain of a host, but the host for the most part doesn't care what the input looks like.

One thought would be to only allow `#requires` in a file, but that won't work in some important scenarios, e.g. `Invoke-Command -File foo.ps1` sends the contents of the file to the remote session, but the remote session doesn't know about a file.

People also use Get-Content or Invoke-WebRequest piped to Invoke-Expression - again losing the fact that the input was really in a file.

Reply:
@lzybkr You're referring to preventing `#requires` in an interactive session here right? I'm not particularly invested in this feature, but saw [PowerShell issue #3803](https://github.com/PowerShell/PowerShell/issues/3803) and implemented [this PR](https://github.com/PowerShell/PowerShell/pull/6469) which still throws an error. Thought I would include it in the RFC for the purposes of discussion.

Reply:
I see - in that context, interactive == add to history. From the code, it's not obvious that it doesn't break the `Invoke-Command` scenario I mentioned, nor is it obvious under what conditions one might invoke something adding to history where the command isn't actually interactive.

Reply:
Ah right! Admittedly part of me thinks that piping into `Invoke-Expression` is something we should discourage, but I suppose we shouldn't break it, and moreover that `#requires` is probably more useful doing what it promises to do there. But good that I now understand the full context of the `interactiveCommand`; I'll roll back a couple of those changes.

Reply:
Seems there is some scenarios where "interactive" is important (ex.: don't load some modules). We already added `-i` to be compatible with Bash. I'd want to get that the engine knows that the session is interactive.
Perhaps @mklement0 have more thoughts.

Reply:
@iSazonov - an interactive session is different than a command entered interactively.

Set-ExecutionPolicy is the only place in the engine that I know of that infers if the command is entered interactively and alters it's behavior accordingly.

Reply:
@iSazonov - an interactive session is different than a command entered interactively.

Set-ExecutionPolicy is the only place in the engine that I know of that infers if the command is entered interactively and alters it's behavior accordingly.